### ISSUE TITLE
Discovery: Investigate fetching web-component library unconditionally for injected header

### ISSUE THREAD

**Issue by ryguyk** [department-of-veterans-affairs/va.gov-cms#11260]
## Description
If a web component is added to the header, the header will break when injected on external TeamSites. The easiest solution to this problem is to fetch the package including all web components every time the header is injected on external sites. We need to investigate the cost of doing so. Is unconditionally fetching the web-component library worth the extra page weight?

### Branch in progress
@ryguyk started a proof of concept in branch, `injected-header-web-components` ([diff](https://github.com/department-of-veterans-affairs/vets-website/compare/main...injected-header-web-components)). Start there, and feel free to talk to Ryan for context.

#### If YES (fetch the web-components package)
If we decide we should fetch this package, it'll have an effect on at least one other ticket currently blocked by this ticket:
1. #11253 The integration test laid out by Brooks Johnson would be a good candidate as a model for what we'd need. If we decide to fetch all the web components, we would need a test that confirms they are all available on the page.
#### If NO (don't fetch the entire web-components package)
If we decide we should not fetch this package, we have two options going forward:
1. Restrict use of web components in the header. We would then possibly look at linting rules to help prevent the use of web components (#11256).
2. Explore the LOE and feasibility of packaging and fetching a subset of web components.

## Acceptance Criteria
- [x] Document performance findings in this ticket:
   - [ ] The increased TeamSite page weight resulting from fetching the entire web-component library 
   - [ ] How increased TeamSite page weight affects various metrics (e.g. First contentful paint, time to interactive, blocking time, etc.)
   - [ ] Comparison to other pages (e.g. va.gov pages that are not external, other accepted standards for page load times, etc.)
- [x] Make a recommendation on whether the entire web-component library should be fetched on all instances of the header being injected.
- [x] Update https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/header-footer/engineering with any learnings about file structure of the header / injected header

**Comment by jilladams** [comment]
@wesrowe @jtmst to connect re: whether there's a math way to derive the report of TeamSite + web component weight, or if some implementation / testing is required, and update estimate after.

**Comment by wesrowe** [comment]
@jtmst and I connected about an approach to estimating the impact of this change that would be more practical than finding an environment where we could try implementing it to measure performance impact. The approach: compare package sizes and assume every kB has the same weight.

Josh, what's the estimate for this approach?

**Comment by jilladams** [comment]
Documenting for posterity: 
* This task was not a great fit for Randi, being so new to the code, and the header being wildly complex in terms of the various files / templates required to run it esp for injected header. Removing that assignment. 
* Randi did get to a POC locally but determined with Ryan that it's not the correct path forward. 
* Ryan also got to a working POC via alternative route that is potentially a better candidate. That work is pushed to `injected-header-web-components`, added to ticket body, and is the right place for the next assignee to pick up. 
  * Ryan did flag that I think the reality is that there are still some unknowns and the POC isn't super high confidence so far, so there's more work to do here to see if it's viable in order to benchmark. 

Putting back into sprint backlog, with goal that Josh or Daniel can pick this up when other sprint work is complete. May ride into next sprint, if we run out of time.

**Comment by jilladams** [comment]
This ticket blocks #11256, which was also in sprint. Have swapped this back in, rather than that 5pt ticket.

**Comment by dsasser** [comment]
# Discovery Findings

Using web components in the injected header/footer was tested using Ryan's branch as a starting point. First I confirmed that this approach is functional. To do this, I added a few randomly selected web components from VA Design System to the header partial and loaded `vets-website` using the `yarn watch local-proxy-rewrite` command. The components rendered as expected in the header (on Desktop, the same wasn't expected on Mobile because of differences in rendering the header for Mobile devices).

## Testing Results

### Size
The overall "weight" of this change in terms of transfer size increased from 2.3mb to 5.5mb. However, this difference can be explained mainly due to the difference in bundle size between production and development. When testing the new injected header w/ web components, the proxy-rewrite-entry.js file size is 4.1MB whereas the production version is 806KB.
![Screenshot 2023-05-11 at 12 28 01 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/a89e7767-1fdb-4209-a1db-5b1f1f988c21)

<details>
  <summary>Size results</summary>

![Screenshot 2023-05-11 at 12 35 21 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/b2706c90-d367-4b74-b00f-5551d37220ea)
![Screenshot 2023-05-11 at 12 35 11 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/2154597b-7466-4bcb-a9cd-b916deb3f27b)


</details>

### Lighthouse
Lighthouse performance score _improved_ by 10 by adding the web components to the injected header.

[new_injected_header-01.json.txt](https://github.com/department-of-veterans-affairs/va.gov-cms/files/11455569/new_injected_header-01.json.txt)
[current_injected_header-01.json.txt](https://github.com/department-of-veterans-affairs/va.gov-cms/files/11455570/current_injected_header-01.json.txt)

<details>
  <summary>Results</summary>

![Screenshot 2023-05-11 at 12 05 14 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/02ffc9ea-7a78-4650-b2a3-ec6e167cf8ff)

</details>

## Testing

To assess the impact of adding web components to the injected header, we took a look at two areas: total size, and Lighthouse performance.

To ensure scores that were less influenced by TeamSite performance variances, testing was performed using a custom minimal static site served by Node locally.

<details>
  <summary>Test server details</summary>
The server was run using the http-server npm package with the command: `npx http-server` (using defaults).

Two HTML5 templates were used to create unique html pages, one for the current injected header test, and one for the new injected header (with web components) test. To each template were added the necessary script and css includes as per the [Platform Documentation on setting up TeamSites](https://depo-platform-documentation.scrollhelp.site/developer-docs/teamsite-overview#TeamSiteoverview-ScriptsandTeamSiteAdministration). Note that the settings.js file was not included for two reasons: the file isn't available at the url in the documentation, and TeamSites aren't using it.

### HTTP Server Setup

<details>
  <summary>Symlink</summary>
 The only requirement outside of running `npx http-server` was to create a symlink to the `content-build` generated output using the command `ln -s ~/projects/content-build/build/localhost/generated .`. This was necessary to allow the proxy-rewrite-entry.js script to fetch assets from the proxied site.

Folder structure as used:
![Screenshot 2023-05-11 at 11 43 07 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/ff8f8a64-5a51-4a42-9b19-a71b10a13641)
</details>

### HTML Templates

<details>
  <summary>For Existing Injected Header Static Page</summary>

```html
<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>A Basic HTML5 Template</title>
  <meta name="description" content="A simple HTML5 Template for new projects.">
  <meta name="author" content="SitePoint">

  <meta property="og:title" content="A Basic HTML5 Template">
  <meta property="og:type" content="website">
  <meta property="og:description" content="A simple HTML5 Template for new projects.">

  <style type="text/css">.brand-consolidation-deprecated { display: none !important; } </style>
  <link rel="stylesheet" href="https://prod-va-gov-assets.s3-us-gov-west-1.amazonaws.com/generated/styleConsolidated.css" />
  <script type="text/javascript" src="https://prod-va-gov-assets.s3-us-gov-west-1.amazonaws.com/generated/polyfills.entry.js"></script>
  <script type="text/javascript" src="https://prod-va-gov-assets.s3-us-gov-west-1.amazonaws.com/generated/vendor.entry.js"></script>
  <script type="text/javascript" src="https://prod-va-gov-assets.s3-us-gov-west-1.amazonaws.com/generated/proxy-rewrite.entry.js"></script>
  <script type="text/javascript" src="https://prod-va-gov-assets.s3-us-gov-west-1.amazonaws.com/js/vendor/uswds.min.js"></script>

</head>

<body>
  <header></header>
  <h1>Current Injected Header</h1>
  <footer>
    <div id="sub-footer"><span class="small-print"></span></div>
  </footer>
</body>
</html>

```

</details>

<details>
  <summary>For New Injected Header Static Page</summary>

```html
<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>A Basic HTML5 Template</title>
  <meta name="description" content="A simple HTML5 Template for new projects.">
  <meta name="author" content="SitePoint">

  <meta property="og:title" content="A Basic HTML5 Template">
  <meta property="og:type" content="website">
  <meta property="og:description" content="A simple HTML5 Template for new projects.">

  <style type="text/css">.brand-consolidation-deprecated { display: none !important; } </style>
  <link rel="stylesheet" href="https://prod-va-gov-assets.s3-us-gov-west-1.amazonaws.com/generated/styleConsolidated.css" />
  <script type="text/javascript" src="https://prod-va-gov-assets.s3-us-gov-west-1.amazonaws.com/generated/polyfills.entry.js"></script>
  <script type="text/javascript" src="https://prod-va-gov-assets.s3-us-gov-west-1.amazonaws.com/generated/vendor.entry.js"></script>
  <script type="text/javascript" src="http://localhost:8080/generated/proxy-rewrite.entry.js"></script>
  <script type="text/javascript" src="https://prod-va-gov-assets.s3-us-gov-west-1.amazonaws.com/js/vendor/uswds.min.js"></script>

</head>

<body>
  <header></header>
  <h1>New Injected Header</h1>
  <footer>
    <div id="sub-footer"><span class="small-print"></span></div>
  </footer>
</body>
</html>

```
</details>

<details>
  <summary>Screenshots</summary>

Static page:
![Screenshot 2023-05-11 at 11 31 22 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/00018b65-ba08-4d16-ba56-cce3f4525d92)

Static page w/ injected header:
![Screenshot 2023-05-11 at 11 32 04 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/f98b456e-42b7-4e68-9157-10e61d3f1da6)

</details>

</details>

### Testing Process

#### Lighthouse
Lighthouse tests were performed using the 'Desktop' type and only the 'Performance' category. Each test was performed 5 times. The least performant of the tests was selected for comparison. Additional notes:

- `vets-website` was launched using `yarn watch local-proxy-rewrite`
- Chrome was started insecurely to avoid CORS issues via the command: `/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --args --user-data-dir="/var/tmp/chrome-dev-disabled-security" --disable-web-security --disable-site-isolation-trials`
- The target url for the test was `http://localhost:3001/?target=http://localhost:8080`.
- Each test was performed in Chrome DevTools. The results were exported to .json files
- The json results were viewed using https://googlechrome.github.io/lighthouse/viewer/
- Once the least performant test was identified for each test type, they were compared using https://googlechrome.github.io/lighthouse-ci/difftool/

#### Size
Total transfer size was determined using Chrome DevTools
![Screenshot 2023-05-11 at 11 58 14 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/621f8a88-cc8a-4cae-ad5f-5a591027eef7)
![Screenshot 2023-05-11 at 12 32 15 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/86b9eea4-9f74-4ac7-9d4a-4acc7883661d)

**Comment by jilladams** [comment]
@dsasser It seems like overall the recommendation / finding is: weight is within an acceptable margin (that'll need signoff) and we recommend moving forward to add the components?

**Comment by dsasser** [comment]
> @dsasser It seems like overall the recommendation / finding is: weight is within an acceptable margin (that'll need signoff) and we recommend moving forward to add the components?

Yes my recommendation would be to move forward with adding the web components to the injected header. Also, my plan is to circle back and add the final recommendation to the discovery output after the team decides how we should move forward.

**Comment by wesrowe** [comment]
@dsasser, I noticed that you ran each version five times and then selected an outlier (worst run) for analysis? All things being equal, I typically see investigators discard outliers and average the remainder. It might not have been practical here, I'm just academically curious about what went into that decision.

**Comment by dsasser** [comment]
> @dsasser, I noticed that you ran each version five times and then selected an outlier (worst run) for analysis? All things being equal, I typically see investigators discard outliers and average the remainder. It might not have been practical here, I'm just academically curious about what went into that decision.

The selected runs were not statistical outliers in my opinion. I chose the worst runs to show worst case scenario rather than present a more optimistic outlook.

**Comment by jilladams** [comment]
Per scrum, no updates to docs needed, so AC is satisfied.

**Comment by jilladams** [comment]
New ticket for implementation will be refined.
